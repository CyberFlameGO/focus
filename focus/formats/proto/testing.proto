syntax = "proto3";

package testing;

message Foo {
	uint32 bar = 1;
}
