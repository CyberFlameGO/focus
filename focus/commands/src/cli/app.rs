use std::sync::Arc;

use crate::{ui::UserInterface, util::sandbox::Sandbox};
use anyhow::{Context, Result};

#[derive(Clone)]
pub struct App {
    ui: Arc<UserInterface>,
    sandbox: Arc<Sandbox>,
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, interactive: bool) -> Result<Self> {
        let ui = Arc::from(UserInterface::new(interactive).context("Failed to start UI")?);
        let sandbox =
            Arc::from(Sandbox::new(preserve_sandbox_contents).context("Failed to create sandbox")?);

        Ok(Self { ui, sandbox })
    }

    /// Get a reference to the app's ui.
    pub fn ui(&self) -> Arc<UserInterface> {
        self.ui.clone()
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }
}
