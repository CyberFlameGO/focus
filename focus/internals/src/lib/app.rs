use std::borrow::Borrow;
use std::sync::Arc;

use crate::{ui::UserInterface, util::sandbox::Sandbox};
use anyhow::{Context, Result};
use std::time::SystemTime;
use ti_library::tool_insights_client::ToolInsightsClient;

#[derive(Clone)]
pub struct App {
    ui: Arc<UserInterface>,
    sandbox: Arc<Sandbox>,
    tool_insights_client: ToolInsightsClient,
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, interactive: bool) -> Result<Self> {
        let ui = Arc::from(UserInterface::new(interactive).context("Failed to start UI")?);
        let sandbox =
            Arc::from(Sandbox::new(preserve_sandbox_contents).context("Failed to create sandbox")?);
        let tool_insights_client = ToolInsightsClient::new(
            // TODO: get this from toml file
            "focus".to_string(),
            // TODO: get this from toml file
            "0.3".to_string(),
            SystemTime::now(),
        );
        Ok(Self {
            ui,
            sandbox,
            tool_insights_client,
        })
    }

    /// Get a reference to the app's ui.
    pub fn ui(&self) -> Arc<UserInterface> {
        self.ui.clone()
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }

    /// Get a reference to the app's tool-insights client.
    pub fn tool_insights_client(&self) -> &ToolInsightsClient {
        self.tool_insights_client.borrow()
    }
}
