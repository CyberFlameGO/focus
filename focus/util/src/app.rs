use std::sync::Arc;
use std::{borrow::Borrow, fmt::Debug};

use crate::sandbox::Sandbox;
use anyhow::{Context, Result};
use std::time::SystemTime;
use tool_insights_client::Client;

#[must_use = "The exit code for the application should be returned and bubbled up to `main` so that it can be passed to `std::process::exit`."]
#[derive(Debug, PartialEq, Eq)]
pub struct ExitCode(pub i32);

#[derive(Clone)]
pub struct App {
    sandbox: Arc<Sandbox>,
    tool_insights_client: Client,
}

impl Debug for App {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("App").finish()
    }
}

impl App {
    pub fn new(preserve_sandbox_contents: bool, with_cmd_prefix: Option<&str>) -> Result<Self> {
        let invocation_description = std::env::args().collect::<Vec<String>>().join(" ");

        let sandbox = Arc::from(
            Sandbox::new(
                Some(&invocation_description),
                preserve_sandbox_contents,
                with_cmd_prefix,
            )
            .context("Failed to create sandbox")?,
        );
        let tool_insights_client = Client::new(
            env!("CARGO_PKG_NAME").to_owned(),
            env!("CARGO_PKG_VERSION").to_owned(),
            SystemTime::now(),
        );
        Ok(Self {
            sandbox,
            tool_insights_client,
        })
    }

    /// Get a reference to the app's sandbox.
    pub fn sandbox(&self) -> Arc<Sandbox> {
        self.sandbox.clone()
    }

    /// Get a reference to the app's tool-insights client.
    pub fn tool_insights_client(&self) -> &Client {
        self.tool_insights_client.borrow()
    }
}
