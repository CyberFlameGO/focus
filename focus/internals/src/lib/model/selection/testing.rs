use std::{collections::HashSet, path::Path, sync::Arc};

use focus_testing::{init_logging, ScratchGitRepo};

use anyhow::Result;
use focus_util::app::App;
use maplit::hashset;
use tempfile::TempDir;

use super::*;

struct Fixture {
    #[allow(dead_code)]
    dir: TempDir,
    underlying: ScratchGitRepo,
    app: Arc<App>,
}

impl Fixture {
    fn new() -> Result<Self> {
        let dir = TempDir::new()?;
        let path = dir.path().join("dense");
        let branch = String::from("main");
        let repo =
            ScratchGitRepo::new_copied_fixture(Path::new("bazel_java_example"), &path, &branch)?;
        let app = Arc::new(App::new(false)?);
        Ok(Self {
            dir,
            underlying: repo,
            app,
        })
    }

    fn repo(&self) -> Result<Repo> {
        Repo::open(self.underlying.path(), self.app.clone())
    }
}

/// Extract project names from the selection
fn project_names(selection: &Selection) -> HashSet<String> {
    selection
        .projects
        .iter()
        .map(|project| project.name.clone())
        .collect()
}

#[test]
fn repo_with_no_selections_returns_mandatory_projects_in_computed_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let selections = repo.selections()?;
    let computed_selection = selections.computed_selection()?;
    assert_eq!(
        project_names(&computed_selection),
        hashset! {String::from("mandatory")}
    );
    assert!(computed_selection.targets.is_empty());

    Ok(())
}

#[test]
fn modifying_and_saving_the_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_name = String::from("team_banzai/project_a");
    let target = Target::try_from("bazel://library_b/...")?;

    {
        let mut selections = repo.selections()?;
        selections.process(&vec![
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Project(project_name.clone()),
            },
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Target(target.clone()),
            },
        ])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});
    }

    {
        // Ensure that after loading from disk in a new instance, the selection is the same.
        let mut selections = repo.selections()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});

        // Remove the target
        selections.process(&vec![Operation {
            action: OperationAction::Remove,
            underlying: Underlying::Target(target),
        }])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name}
        );
        assert!(computed_selection.targets.is_empty());
    }

    Ok(())
}

#[test]
fn adding_an_unknown_project() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let nonexistent_project = Underlying::Project(String::from("blofeld/moonbase"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: nonexistent_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {nonexistent_project});

    Ok(())
}

#[test]
fn mandatory_projects_cannot_be_selected() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let mandatory_project = Underlying::Project(String::from("mandatory"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: mandatory_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {mandatory_project});

    Ok(())
}

#[test]
fn duplicate_projects_are_ignored() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_b = Underlying::Project(String::from("team_zissou/project_b"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.added, hashset! {project_b.clone()});

    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.ignored, hashset! {project_b});

    Ok(())
}
