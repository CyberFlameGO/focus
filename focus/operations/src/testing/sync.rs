use std::{collections::HashSet, path::Path};

use anyhow::Result;
use maplit::hashset;

use focus_testing::init_logging;
use focus_util::app;

use crate::testing::integration::{RepoDisposition, RepoPairFixture};

#[test]
fn sync_upstream_changes() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;

    fixture.perform_clone()?;

    // Commit new files affecting the build graph to the dense repo
    let build_bazel_content = r#"filegroup(
name = "excerpts",
srcs = [
    "catz.txt",
],
visibility = [
    "//visibility:public",
],
)"#;
    fixture.dense_repo.write_and_commit_file(
        Path::new("x/BUILD.bazel"),
        build_bazel_content.as_bytes(),
        "Add excerpts",
    )?;
    let catz_txt_content = r#"The Naming of Cats is a difficult matter,
It isn't just one of your holiday games
    )"#;
    fixture.dense_repo.write_and_commit_file(
        Path::new("x/catz.txt"),
        catz_txt_content.as_bytes(),
        "Add excerpts",
    )?;

    // Fetch in the sparse repo from the dense repo
    fixture.perform_pull(RepoDisposition::Sparse, "origin", "main")?;

    // Make sure that the graph is seen as having changed
    assert_eq!(
        crate::detect_build_graph_changes::run(
            &fixture.sparse_repo_path,
            vec![],
            fixture.app.clone(),
        )?,
        app::ExitCode(1)
    );

    // Sync in the sparse repo
    crate::sync::run(&fixture.sparse_repo_path, false, fixture.app.clone())?;

    let x_dir = fixture.sparse_repo_path.join("x");
    assert!(!x_dir.is_dir());

    // Add as a target
    crate::selection::add(
        &fixture.sparse_repo_path,
        true,
        vec![String::from("bazel://x/...")],
        fixture.app.clone(),
    )?;

    assert!(x_dir.is_dir());

    Ok(())
}

#[test]
fn sync_layer_manipulation() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let selected_project_names = || -> Result<HashSet<String>> {
        Ok(fixture
            .sparse_repo()?
            .selection_manager()?
            .computed_selection()?
            .projects
            .iter()
            .filter_map(|project| {
                if project.mandatory {
                    None
                } else {
                    Some(project.name.to_owned())
                }
            })
            .collect::<HashSet<String>>())
    };

    let project_a_label = String::from("team_banzai/project_a");
    let project_b_label = String::from("team_zissou/project_b");

    let path = fixture.sparse_repo_path.clone();
    let library_a_dir = path.join("library_a");
    let project_a_dir = path.join("project_a");
    let library_b_dir = path.join("library_b");
    let project_b_dir = path.join("project_b");
    let profile_path = path.join(".git").join("info").join("sparse-checkout");

    {
        let selected_names = selected_project_names()?;
        assert_eq!(selected_names, hashset! {});
    }
    insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);

    assert!(!library_b_dir.is_dir());
    assert!(!project_b_dir.is_dir());
    crate::selection::add(
        &path,
        true,
        vec![project_b_label.clone()],
        fixture.app.clone(),
    )?;
    {
        let selected_names = selected_project_names()?;
        assert_eq!(selected_names, hashset! { project_b_label.clone() })
    }

    insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
    assert!(library_b_dir.is_dir());
    assert!(project_b_dir.is_dir());

    assert!(!library_a_dir.is_dir());
    assert!(!project_a_dir.is_dir());
    crate::selection::add(
        &path,
        true,
        vec![project_a_label.clone()],
        fixture.app.clone(),
    )?;
    {
        let selected_names = selected_project_names()?;
        assert_eq!(
            selected_names,
            hashset! { project_a_label.clone(), project_b_label.clone() }
        )
    }
    insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
    assert!(library_a_dir.is_dir());
    assert!(project_a_dir.is_dir());

    crate::selection::remove(&path, true, vec![project_a_label], fixture.app.clone())?;
    {
        let selected_names = selected_project_names()?;
        assert_eq!(selected_names, hashset! { project_b_label.clone() })
    }
    insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);
    assert!(!library_a_dir.is_dir());
    assert!(!project_a_dir.is_dir());

    crate::selection::remove(&path, true, vec![project_b_label], fixture.app.clone())?;
    {
        let selected_names = selected_project_names()?;
        assert_eq!(selected_names, hashset! {});
    }
    insta::assert_snapshot!(std::fs::read_to_string(&profile_path)?);

    assert!(!library_b_dir.is_dir());
    assert!(!project_b_dir.is_dir());

    Ok(())
}

#[test]
fn sync_adhoc_manipulation() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let path = fixture.sparse_repo_path.clone();
    let library_b_dir = path.join("library_b");
    let targets = vec![String::from("bazel://library_b/...")];

    crate::selection::add(
        &fixture.sparse_repo_path,
        true,
        targets.clone(),
        fixture.app.clone(),
    )?;
    assert!(library_b_dir.is_dir());

    crate::selection::remove(
        &fixture.sparse_repo_path,
        true,
        targets.clone(),
        fixture.app.clone(),
    )?;
    assert!(!library_b_dir.is_dir());

    Ok(())
}

#[test]
fn failed_selection_mutations_are_reverted() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let selections = fixture.sparse_repo()?.selection_manager()?;
    let selection_before = selections.selection()?;
    let targets = vec![String::from("bazel://library_z/...")];
    assert!(crate::selection::add(
        &fixture.sparse_repo_path,
        true,
        targets,
        fixture.app.clone()
    )
    .is_err());
    let selection_after = selections.selection()?;
    assert_eq!(selection_before, selection_after);

    Ok(())
}

#[test]
fn clone_contains_top_level() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let sparse_repo = fixture.sparse_repo()?;
    let outlining_tree = sparse_repo.outlining_tree().unwrap();
    let underlying = outlining_tree.underlying();
    let outlining_tree_root = underlying.work_dir();

    let top_level_bazelisk_rc = outlining_tree_root.join(".bazeliskrc");

    let outlining_tree_git_dir = sparse_repo
        .git_dir()
        .join("worktrees")
        .join("outlining-tree");

    let profile =
        std::fs::read_to_string(outlining_tree_git_dir.join("info").join("sparse-checkout"))?;
    insta::assert_snapshot!(&profile);

    assert!(top_level_bazelisk_rc.is_file());

    Ok(())
}

#[test]
fn sync_skips_checkout_with_unchanged_profile() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let path = fixture.sparse_repo_path.clone();
    let targets = vec![String::from("bazel://library_b/...")];
    crate::selection::add(
        &fixture.sparse_repo_path,
        false, // Note: Manual sync
        targets.clone(),
        fixture.app.clone(),
    )?;
    // First sync performs a checkout.
    assert!(crate::sync::run(&path, false, fixture.app.clone())?.checked_out);

    // Subsequent sync does not perform a checkout.
    assert!(!crate::sync::run(&path, false, fixture.app.clone())?.checked_out);

    Ok(())
}

#[test]
fn sync_configures_working_and_outlining_trees() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    // Check working tree
    let working_tree_config = fixture.sparse_repo()?.underlying().config()?.snapshot()?;
    assert!(working_tree_config.get_bool("index.sparse")?);
    assert!(working_tree_config.get_bool("core.untrackedCache")?);

    // Check outlining tree
    let outlining_tree_config = fixture
        .sparse_repo()?
        .outlining_tree()
        .unwrap()
        .underlying()
        .git_repo()
        .config()?
        .snapshot()?;
    assert!(outlining_tree_config.get_bool("index.sparse")?);
    assert!(outlining_tree_config.get_bool("core.untrackedCache")?);

    Ok(())
}

#[test]
fn regression_adding_directory_targets_present_in_mandatory_sets() -> Result<()> {
    init_logging();

    let fixture = RepoPairFixture::new()?;
    fixture.perform_clone()?;

    let path = fixture.sparse_repo_path.clone();

    let mandatory_y_dir = path.join("mandatory_y");
    let very_important_info_dir = mandatory_y_dir.join("very_important_info");
    let automakers_dir = very_important_info_dir.join("automakers");
    let swedish_txt_file = automakers_dir.join("swedish.txt");
    let targets = vec![String::from("directory:mandatory_y")];
    assert!(automakers_dir.is_dir());
    assert!(swedish_txt_file.is_file());
    crate::selection::add(
        &fixture.sparse_repo_path,
        true,
        targets.clone(),
        fixture.app.clone(),
    )?;
    assert!(swedish_txt_file.is_file());

    Ok(())
}
