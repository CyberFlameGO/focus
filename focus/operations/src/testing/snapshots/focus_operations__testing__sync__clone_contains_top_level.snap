---
source: focus/operations/src/testing/sync.rs
expression: "&profile"
---
/*
!/*/
/focus/
*.bzl
BUILD*
WORKSPACE*

